### ISSUE TITLE
Json parsing error behaviors control in Minimal Api

### ISSUE THREAD

**Issue by ziaulhasanhamim** [dotnet/aspnetcore#44538]
## Background and Motivation

I was porting one of my API from controllers to minimal API. I noticed an unexpected behavior in minimal API. Like when I have invalid JSON in the request body in controllers the parameter binds to null and there is an error message added to `ModelStateDictionary`. But in minimal APIs, it throws a JsonException, most importantly the request doesn't even reach the handler method. I wasn't expecting that. Because in my mind minimal API suppose to give me more flexibility. 

## Proposed API

There can be a simple solution. Instead of throwing a JsonException why not just provide the handler with correct error messages?

```c#
public class ParsingError
{
   // necessary infors
}
public class ParsingErrorParamaterAttribute : Attribute
{
   public ParsingErrorParamaterAttribute(string paramName) {}
}
```

Here `ParsingError` will contain whole contexts about the wrong request body format. So for this to happen with custom bindings there needs to be a new overload of the BindAsync method. 

```c#

public class Post
{
    public static ValueTask<BindingResult<T>> BindAsync(HttpContext context, ParameterInfo parameter)
    {
        // binding code
    }
}
public record BindingResult<T>(T? Model, ParsingError? Error);
```

So now custom bindings can also return ParsingError. 

## Usage Examples

The usage would be very simple. There will be just one extra parameter.
```C#
app.MapPost("api/posts/",  (Post post, [ParsingErrorParamater("post")] ParsingError postParsingError) => 
{
  // handler code
});
```

## Alternative Designs

There can be a configuration option to turn this on or off.

**Comment by davidfowl** [comment]
Yes, we've discussed this situation many times, but model state is *expensive*. I like the idea of a generic type that can capture the appropriate. What you have above doesn't quite work as well because a valid, non-null Post object still needs to be provided to the method. Instead, maybe we can have something like this:

```C#
app.MapPost("api/posts/",  (ModelState<Post> modelState) => 
{
    if (!modelState.IsValid)
    {
        return Results.Problem(modelState.Errors);
    }
    var post = modelState.Model;

    // Do stuff here ...
});
```

**Comment by ziaulhasanhamim** [comment]
Yeah, that approach is much cleaner.

> we've discussed this situation many times, but the model state is expensive

The model state would be expensive because of validations. But this would be inappropriate to call it ModelState.  It would just contain the errors regarding the parsing of JSON and nothing else. So I don't it would be expensive to use `BindingResult<Post>` over `Post`. It would rather simplify request body parsing.

**Comment by davidfowl** [comment]
Correct, if it was simple like the above, then we can have a well-known type/shape for preserving the error that you can opt-into in your handler.

**Comment by ziaulhasanhamim** [comment]
So is there any plan to add a type like this `Parsable<T>` or `BindingResult<T>` in ASP.NET? Because I wouldn't like idea of giving default validation provider in minimal apis. I like the current approach where I can plug in any validators. But for json errors `BindingResult<T>` is useful.

**Comment by davidfowl** [comment]
Well you filed an issue, that's the first step. The team will see this issue and make a decision 😄. It's still the weekend.

**Comment by rellis-of-rhindleton** [comment]
This issue is approaching a year old, but it's still open and I'm having similar problems, so I'll chime in with some context.

(BTW: Using the approach detailed above seems to break OpenAPI for that endpoint, which is a dealbreaker. Am I missing something?)

What I need is the ability to control the error message and make it useful. I'm not in love with ModelState but it allowed me to do that for bad request errors fairly easily, whether from within the handler method or in other ways. I'm not finding a way to do this with minimal APIs.

Say I omit a required property from an input model. A BadHttpRequestException is thrown:

```Failed to read parameter "ModelType model" from the request body as JSON.```

This message doesn't tell me what was wrong with the request. It also shares internal names, and it sounds like the problem happened on the server side, which is misleading.

The inner exception is a JsonException and looks like this:

```JSON deserialization for type 'ModelType' was missing required properties, including the following: propertyName```

This message contains internal information which should not be shared, but it also contains information that *should* be shared. I'd like to tell the client about missing required properties and which ones they are as part of a ProblemDetails response. But I can't get to that without scraping the string, which doesn't seem like a good idea. This seems to be a problem with System.Text.Json's JsonException and ThrowHelper, which starts out with the information I want but doesn't keep it separate. There may not be anything you can do about that.

**Comment by gorillapower** [comment]
Any ideas if this is on the radar for .NET 10?

**Comment by yuxingz** [comment]
in `Production` Environment，`ThrowOnBadRequest` default `false` ，Set it to true will solve it.
```csharp
        services.Configure<RouteHandlerOptions>(options =>
        {
            options.ThrowOnBadRequest = true;
        });
```
